pkg/runtime/depot.go

```diff
@@ -300,6 +331,46 @@ func (d *depot) validateVolume(absoluteDest string) error {
 	return nil
 }
 
+// getSharedFilesToReLink returns a map of deployed files to re-link to (or re-copy from) another
+// artifact that provides those files. The key is the deployed file path and the value is the
+// source path from another artifact.
+func (d *depot) getSharedFilesToReLink(id strfmt.UUID, relativeSrc, path string) (map[string]string, error) {
+	// Map of deployed paths to other sources that provides those paths.
+	relink := make(map[string]string, 0)
+
+	// Get a listing of all files deployed by this artifact.
+	deployedDir := filepath.Join(d.Path(id), relativeSrc)
+	deployedFiles, err := fileutils.ListDirSimple(deployedDir, false)
```

Comment:
We should calculate this purely based on the deployment config. Since we store the installed files it should have all the information we need.

Main point here is that the config is our source of truth, and if the on-disk data doesn't agree with that then we have  an issue that we'll need to raise. If we inconsistently use a different source of truth these types of conflicts will be difficult to diagnose.

Additionally, the deployed dir is shared between multiple artifacts. So the listing you get here isn't representative for the given artifact.

Reply:
Good points!